jwt/src/main/scala/it/pagopa/interop/commons/jwt/service/JWTReader.scala

```diff
@@ -1,102 +1,65 @@
-        optionalHeaderValueByName("Authorization").flatMap {
-          case Some(header) =>
-            header.split(" ").toList match {
-              case "Bearer" :: payload :: Nil =>
-                validation.andThen(authenticationDirective(contextInfo))(payload)
-              case _                          =>
-                logger.warn(s"$contextInfo - No authentication has been provided for this call")
-                reject(MalformedHeaderRejection("Authorization", "Illegal header key."))
-            }
-          case None         =>
-            logger.warn(s"$contextInfo - No authentication has been provided for this call")
-            reject(AuthenticationFailedRejection(CredentialsMissing, HttpChallenge("Bearer", None)))
+  def OAuth2JWTValidatorAsContexts(implicit
+    logger: LoggerTakingImplicit[ContextFieldsToLog]
+  ): Directive1[Seq[(String, String)]] = for {
+    req       <- extractRequest
+    ctx       <- optionalHeaderValueByName(CORRELATION_ID_HEADER).map(
+      _.fold(Seq.empty[(String, String)])(cid => Seq(CORRELATION_ID_HEADER -> cid))
+    )
+    maybeAuth <- optionalHeaderValueByName("Authorization").map(_.map(_.split(" ").toList))
+    res       <- maybeAuth match {
+      case Some("Bearer" :: payload :: Nil) =>
+        bearerAsContexts(payload) match {
+          case Success(x)  => provide(x)
+          case Failure(ex) =>
+            logger.warn(s"Invalid authentication provided - ${ex.getMessage}")(ctx)
+            reject(AuthenticationFailedRejection(CredentialsRejected, HttpChallenge("Bearer", None)))
+              .toDirective[Tuple1[Seq[(String, String)]]]
         }
-      }
+      case Some(_)                          =>
+        logger.warn(s"No authentication has been provided for this call ${req.method.value} ${req.uri}")(ctx)
+        reject(MalformedHeaderRejection("Authorization", "Illegal header key."))
+          .toDirective[Tuple1[Seq[(String, String)]]]
+      case None                             =>
+        logger.warn(s"No authentication has been provided for this call ${req.method.value} ${req.uri}")(ctx)
+        reject(AuthenticationFailedRejection(CredentialsMissing, HttpChallenge("Bearer", None)))
+          .toDirective[Tuple1[Seq[(String, String)]]]
     }
-  }
+  } yield res
 
-  private def authenticationDirective[T](contextInfo: String): Try[T] => Directive1[T] = {
-    case Success(result) => provide(result)
-    case Failure(ex)     =>
-      logger.warn(s"$contextInfo - Invalid authentication provided - ${ex.getMessage}")
-      reject(AuthenticationFailedRejection(CredentialsRejected, HttpChallenge("Bearer", None)))
+  private def bearerAsContexts(bearer: String): Try[Seq[(String, String)]] = for {
+    claims         <- getClaims(bearer)
+    uid            <- Try(Option(claims.getStringClaim(UID)).getOrElse(""))
+    sub            <- Try(Option(claims.getSubject).getOrElse(""))
+    organizationId <- Try(claims.getStringClaim(ORGANIZATION_ID_CLAIM))
```

Comment:
may I suggest to handle the nullability right here to avoid carrying around a possible `null`?
```suggestion
    organizationId <- Try(Option(claims.getStringClaim(ORGANIZATION_ID_CLAIM)).getOrElse(""))
```

Reply:
I totally can, but I have to follow the stack of calls to see whether it changes some behaviour downstream :D

Reply:
btw it's handled a few lines below

```scala
val orgId: List[(String, String)] = Option(organizationId).map(ORGANIZATION_ID_CLAIM -> _).toList
```

but it's not added to the context list. I can't tell whether changing this behaviour will break something downstream, I prefer to leave it as it is in this PR. Do you agree?

Reply:
Not sure I got your point, the `organizationId` would be handled the same way, the only other place where it is used is where you indicated, and it's still handled as an `Option`.
What I'm suggesting is just to wrap it as soon as it is obtained from the claim to avoid future changes to mistakenly use a `null` value